src/callbacks/callback_imcomm.cpp

```diff
@@ -132,16 +132,20 @@ void lbann_callback_imcomm::on_epoch_end(model *m) {
     optimizer *opt = w->get_optimizer();
     if (ct_does_quantization(params.ct)) {
       comm->intermodel_sum_matrix(params.error);
+      opt->clear_gradient();
+      auto gradient = opt->get_gradient().Copy();
       Mat *local_gradients = nullptr;
       Mat reshaped;
       if (params.reshape_height > 0) {
-        reshape_mat(opt->get_gradient().Matrix(),
+        reshape_mat(gradient->Matrix(),
                     reshaped, params.reshape_height, params.reshape_width);
         local_gradients = &reshaped;
       } else {
-        local_gradients = &(opt->get_gradient().Matrix());
+        local_gradients = &(gradient->Matrix());
       }
       *local_gradients = params.error;
+      opt->add_to_gradient(*gradient);
```

Comment:
The old implementation did not properly set the gradient if the data was on GPU.